### PR TITLE
Neat initial commit

### PR DESCRIPTION
**Pull request by Vezzp**
Перенёс модели, которые лежали в `models` в `sslearn/models`, в папки `rf` и `lda` соответственно.

### CODE REVIEW

**Comment by BorisovDm** [comment]
Может быть логи не сохранять?
Просто вынести пример вывода в README?

File: sslearn/datasets.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    @property
+    def all_ldata(self) -> np.ndarray:
+        if len(self.pseudo_ldata):
+            return np.vstack([self.ldata, self.pseudo_ldata])
+        return self.ldata
+
+    @property
+    def udata(self) -> np.ndarray:
+        return self._udata
```
Comment (BorisovDm) [inline sslearn/datasets.py:43]: Ты вроде бы хранишь self.udata?

File: sslearn/datasets.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        self.ldata = ldata
+        self.labels = labels
+        self.udata = udata
+
+    @property
+    def all_labels(self) -> np.ndarray:
+        if len(self.pseudo_labels):
+            return np.hstack([self.labels, self.pseudo_labels])
+        return self.labels
```
Comment (BorisovDm) [inline sslearn/datasets.py:33]: Это нельзя вызвать сразу после init?
Тогда self.pseudo_labels будет не определено

File: sslearn/models/clustering/ssclustering.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
+import numpy as np
+from scipy.spatial.distance import pdist, squareform
+
+
+class SSClustering(object):
+    def __init__(self, eps: float, min_points: int):
+        """
+        """
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:8]: Может не нужно без текста

File: sslearn/models/clustering/ssclustering.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        self.eps = eps  # eps is the radius of the ball in which we look for neighbours of the point
+        self.min_points = min_points  # min_points is the minimum number of points required to form a dense region
+        self.n_clusters = None
+
+    def fit(self, ldata, labels, udata):
+        """
+        """
+        # DF is a m x n matrix, where m is number of data for training and n is the number of features, it it x_u_train
+        DF, x_l_train, y_l_train = udata, ldata, labels
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:17]: Maybe lowercase?

File: sslearn/models/clustering/ssclustering.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        # Initialization
+        DF_merged = np.concatenate((DF, x_l_train))
+        distance = squareform(pdist(DF_merged))
+        # The square matrix with distances from each point to each point
+        # print(np.mean(distance))
+        m = DF_merged.shape[
+            0
+        ]  # the amount of data to be divided into clusters (number of rows in DF)
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:25]: Может быть всё-таки в одну строку?
И еще пробелов добавить между логическими кусками

File: sslearn/models/clustering/ssclustering.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            m
+        )  # the array: if 0 then point is border of cluster, if 1 then point is the core point of the cluster
+        # if 2 then point is noise
+        cluster = []  # the array of points of certain cluster
+        clusters = []  # The list of clusters
+        point_cluster_index = np.zeros(
+            m
+        )  # the array containing the cluster number of each point
+        cluster_index = 0  # cluster counter
+        Current_neighbors = []
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:40]: сurrent_neighbors

File: sslearn/models/clustering/ssclustering.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                            if len(neighbors) >= self.min_points:
+                                neighbors_list = []
+                                for p in neighbors:
+                                    neighbors_list.append(p.tolist())
+                                for q in range(len(neighbors_list)):
+                                    if neighbors_list[q] not in Current_neighbors_list:
+                                        Current_neighbors_list.append(neighbors_list[q])
+                        if point_cluster_index[b] == 0:
+                            cluster.append(b)
+                            point_cluster_index[b] = cluster_index
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:70]: Там в алгоритме реально такая большая вложенность?
Нельзя как-то упростить?

File: sslearn/models/clustering/ssclustering.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+                    cluster.append(Current_neighbors_list[:])
+                    clusters.append(cluster[:])
+        # print(cluster_index)
+        self.cluster_index = cluster_index
+        self.point_cluster_index = point_cluster_index
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:76]: Может сразу с self переменными работать?

File: sslearn/models/clustering/ssclustering.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            if i == 0:
+                chan[i] = -1
+            elif i in res:
+                k = 0
+                for j in range(len(res)):
+                    if int(res[j] == i):
+                        if int(y_l_train[j]) == 0:
+                            k = k - 1
+                        elif int(y_l_train[j]) == 1:
+                            k = k + 1
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:93]: Может через zip итерироваться по res и y_l_train?

File: sslearn/models/clustering/ssclustering.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        self.point_cluster_index = point_cluster_index
+
+        m_u = DF.shape[0]
+        m_l = x_l_train.shape[0]
+        res = np.zeros(m_l)
+        res = point_cluster_index[m_u:m]
+        chan = np.zeros(cluster_index + 1)
+        for i in range(0, cluster_index + 1):
+            if i == 0:
+                chan[i] = -1
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:85]: Может инициализировать сразу первый элемент(или весь массив) -1?

File: sslearn/models/clustering/ssclustering.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+            else:
+                chan[i] = -1
+        for i in range(0, cluster_index + 1):
+            point_cluster_index[point_cluster_index == i] = chan[i]
+        x_new = []
+        for i in range(0, m):
+            if (point_cluster_index[i] == 0) or (point_cluster_index[i] == 1):
+                x_new.append(i)
+        self.point_cluster_index = point_cluster_index
+        self.x_new = x_new
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:109]: Тоже про self

**Comment by Vezzp** [comment]
Всё, что находится в `ssclustering`, — код Насти, немного отформатированный мной.
Я не вникал в суть, но соглашусь, что можно сделать гораздо чище.

File: sslearn/models/clustering/ssclustering.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+        X_train_1 = []
```
Comment (BorisovDm) [inline sslearn/models/clustering/ssclustering.py:111]: Можно попробовать так:
X_train_1 = np.array([DF_merged[x] for x in x_new])

File: sslearn/models/sla/sla.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        for gamma in gammas:
+            # START: Calculate joint bayes risk, eq. (6).
+            mask_theta_leq = margins <= theta
+            mean_theta_leq = np.mean(margins * mask_theta_leq)
+
+            mask_gamma_l = margins < gamma
+            mean_gamma_l = np.mean(margins * mask_gamma_l)
+
+            # The first term, which is Gibbs risk, is just an estimation.
+            k_u = 0.5 + 0.5 * (np.mean(margins) - 1)
```
Comment (BorisovDm) [inline sslearn/models/sla/sla.py]: k_u = np.mean(margins) / 2?

File: sslearn/models/sla/sla.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+    # Optimize joint bayes risk.
+    for theta in thetas:
+        for gamma in gammas:
+            # START: Calculate joint bayes risk, eq. (6).
+            mask_theta_leq = margins <= theta
+            mean_theta_leq = np.mean(margins * mask_theta_leq)
+
+            mask_gamma_l = margins < gamma
+            mean_gamma_l = np.mean(margins * mask_gamma_l)
```
Comment (BorisovDm) [inline sslearn/models/sla/sla.py:62]: Одна переменная сначала bool, а потом float? Не очень

File: test/dataloaders.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
+import os
+
+from collections import OrderedDict
+
+import numpy as np
+import pandas as pd
+
+from sklearn.preprocessing import LabelEncoder
+from sklearn.datasets import load_svmlight_file
```
Comment (BorisovDm) [inline test/dataloaders.py]: Ну если вспомнить PEP8, то импорты неправильно стоят(=

File: sslearn/datasets.py (hunk 2 above)
Comment (Vezzp) [reply sslearn/datasets.py:33]: А зачем это вызывать после него? Так `self.pseudo_labels` создаются в `__init__`, `self.udata` — это property.

File: test/dataloaders.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+        "telescope": "08_magic_gamma_telescope",
+        "pendigits_4_9": "Pendigits",
+        "mnist_4_9": "MNIST"
+    }
+)
+
+
+def binarize_pendigits(root: str, *labels):
+    if len(labels) != 2:
+        raise ValueError(f"Accepts only two numbers, got {len(labels)}.")
```
Comment (BorisovDm) [inline test/dataloaders.py]: true метод такое делать через assert, ну если тебе не нужен специфичный exception

File: sslearn/models/clustering/ssclustering.py (hunk 3 above)
Comment (Vezzp) [reply sslearn/models/clustering/ssclustering.py:8]: Там должен был быть docstring.

File: sslearn/models/sla/sla.py (hunk 14 above)
Comment (Vezzp) [reply sslearn/models/sla/sla.py:62]: Согласен.

Comment (Vezzp) [reply sslearn/models/sla/sla.py:62]: Хотя нет, это разные же переменные.:)

File: test/dataloaders.py (hunk 15 above)
Comment (Vezzp) [reply test/dataloaders.py]: Что-что, а импорты я точно делаю не по PEP8.

File: sslearn/datasets.py (hunk 1 above)
Comment (Vezzp) [reply sslearn/datasets.py:43]: `self.udata` — это property. В реальности данные хранятся в `self._data`.

File: sslearn/datasets.py (hunk 2 above)
Comment (BorisovDm) [reply sslearn/datasets.py:33]: Так self.pseudo_labels создаются в __init__ - нет

File: test/dataloaders.py (hunk 15 above)
Comment (BorisovDm) [reply test/dataloaders.py]: У меня просто есть болезненный опыт раставления импортов по PEP8)
Хотя я уже забыл их порядок

File: README.md [hunk 17]
```diff
@@ -1,18 +1,64 @@
...
+import sys
+
+repo_path = "SSL"
+sys.path.append(repo_path)
+
+from experimentarium.utils import ShellTestRunner  # noqa
+
+
+if __name__ == "__main__":
+    run_root = "SSL/experimentarium"
```
Comment (BorisovDm) [inline README.md:43]: @Vezzp 
Смог запустить только при run_root = "experimentarium"

File: README.md [hunk 18]
```diff
@@ -1,18 +1,64 @@
...
+        "n_states": 1,
+        "log": "True"
+    }
+    ShellTestRunner().run(run_root, **run_params)
+```
+
+The only thing now is to change `run_params` to whatever you need and execute:
+
+```bash
+python3 ../SSL/run.py
```
Comment (BorisovDm) [inline README.md:62]: А почему не `python3 run.py` ?

Comment (Vezzp) [reply README.md:62]: Я предполагал, что мы находимся в склонированном репозитории, тогда путь до этого файла действительно такой.

File: README.md (hunk 17 above)
Comment (Vezzp) [reply README.md:43]: Странно, у меня работает при `SSL/experimentarium`, может, ты положил `run.py` в папку `SSL`, то есть внутрь репозитория, в то время как я положил его в одной директории с `SSL`?

File: test/dataloaders.py (hunk 16 above)
Comment (Vezzp) [reply test/dataloaders.py]: Насколько я понимаю, `assert` предназначен для тех мест, где падать не должно, но при некоторых обстоятельствах может, но пользователю об этом знать не нужно. То есть когда падает `assert` — обычно означает что-то плохое. Другое дело, что тут излишне использовать `*labels`, так как всего лишь два числа ожидаются. Пожалуй, это нужно исправить.

File: test/dataloaders.py (hunk 15 above)
Comment (Vezzp) [reply test/dataloaders.py]: Я примерно так расставляю. Сначала идут импорты стандартной библиотеки, потом идут сторонние библиотеки, написанные не мной, а потом уже идут мои библиотеки. Каждую из этих трёх групп я разделяю пустой строчкой, секции `import` и `from ... import` я так же разделяю пустой строчкой.

File: sslearn/models/sla/sla.py (hunk 13 above)
Comment (Vezzp) [reply sslearn/models/sla/sla.py]: Да, ты прав, но я написал так намеренно, чтобы себе напомнить о том, что я считаю не вполне точно по алгоритму, а использую верхнюю оценку в 0.5 для риска Гиббса. Пожалуй, можно просто расширить комментарий на этот счёт и использовать `k_u = np.mean(margins) / 2`.

File: README.md (hunk 18 above)
Comment (BorisovDm) [reply README.md:62]: `Я предполагал, что мы находимся в склонированном репозитории` - то есть в папке SSL? Затем мы поднимаемся вверх, а потом снова заходим в SSL?

File: README.md (hunk 17 above)
Comment (BorisovDm) [reply README.md:43]: Да, я создал `run.py` в папке SSL. Просто из Readme мне показалось, что нужно так сделать:
```
There is also a possibility to avoid typing arguments to variety command line arguments, just create file,, for instance,../SSL/run.py with the following content.
```

Comment (Vezzp) [reply README.md:43]: Окей. Будет здорово, если ты сформулируешь понятнее.

**Comment by BorisovDm** [comment]
> Добавил.

Но коммит не сделал?
